pkg/errors/errors.go

```diff
@@ -8,59 +8,170 @@ import (
+	UpdateFailed                error = NewMocError(moccodes.UpdateFailed)
+	NotInitialized              error = NewMocError(moccodes.NotInitialized)
+	NotImplemented              error = NewMocError(moccodes.NotImplemented)
+	OutOfRange                  error = NewMocError(moccodes.OutOfRange)
+	AlreadySet                  error = NewMocError(moccodes.AlreadySet)
+	NotSet                      error = NewMocError(moccodes.NotSet)
+	InconsistentState           error = NewMocError(moccodes.InconsistentState)
+	PendingState                error = NewMocError(moccodes.PendingState)
+	WrongHost                   error = NewMocError(moccodes.WrongHost)
+	PoolFull                    error = NewMocError(moccodes.PoolFull)
+	NoActionTaken               error = NewMocError(moccodes.NoActionTaken)
+	Expired                     error = NewMocError(moccodes.Expired)
+	Revoked                     error = NewMocError(moccodes.Revoked)
+	Timeout                     error = NewMocError(moccodes.Timeout)
+	RunCommandFailed            error = NewMocError(moccodes.RunCommandFailed)
+	InvalidToken                error = NewMocError(moccodes.InvalidToken)
+	Unknown                     error = NewMocError(moccodes.Unknown)
+	DeleteFailed                error = NewMocError(moccodes.DeleteFailed)
+	DeletePending               error = NewMocError(moccodes.DeletePending)
+	FileNotFound                error = NewMocError(moccodes.FileNotFound)
+	PathNotFound                error = NewMocError(moccodes.PathNotFound)
+	NotEnoughSpace              error = NewMocError(moccodes.NotEnoughSpace)
+	AccessDenied                error = NewMocError(moccodes.AccessDenied)
+	BlobNotFound                error = NewMocError(moccodes.BlobNotFound)
+	GenericFailure              error = NewMocError(moccodes.GenericFailure)
+	NoAuthenticationInformation error = NewMocError(moccodes.NoAuthenticationInformation)
+	MeasurementUnitError        error = NewMocError(moccodes.MeasurementUnitError)
+	QuotaViolation              error = NewMocError(moccodes.QuotaViolation)
+	IPOutOfRange                error = NewMocError(moccodes.IPOutOfRange)
+	MultipleErrors              error = NewMocError(moccodes.MultipleErrors)
 )
 
+// GetMocErrorCode attempts to extract the MocCode from the given error. If the error is a multierr,
+// it will return the MocCode if all errors in the multierr match the same MocCode. If there are
+// multiple non-matching MocCodes, then it returns MocCode.MultipleErrors and false.
+//
+// GetMocErrorCode follows the following rules when parsing individual errors:
+//
+//   - If the error is nil, it returns MocCode.OK and true.
+//   - If the error is of type MocError, it returns the MocCode and true.
+//   - If the error has a Cause that is not nil and is of type MocError, it returns the MocCode of the Cause and true.
+//   - If both the error and its Cause are not of type MocError, it returns MocCode.Unknown and false.
+func GetMocErrorCode(err error) (moccodes.MocCode, bool) {
```

Comment:
It might be difficult to review these functions without seeing the usage in the agent code. I wonder if it makes sense to merge this PR after the agent PRs are reviewed as well.

Reply:
Fair point. Most of the functions are currently only used internally to the pkg. I think the ones I currently use in the agent PRs are:
1. NewMocErrorWithError: https://github.com/microsoft/wssdcloudagent/blob/08bd4b8732526eec308e705f263f3c5bdd73add7/pkg/cloudagent/services/storage/virtualharddisk/wssd/wssd.go#L288
2. IsMocErrorCode (for VHD decide to block detach or not)